[14.0][IMP] cetmix_tower_server_queue-async_file_operations

Kurama2d opened the pull request:
Use queue_job to process file upload/download in background.
Add is_processed field and UI indicators for status display.

Task: 3720


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced asynchronous file upload and download capabilities managed via a job queue, including processing state tracking and user notifications.
  - Enhanced file views with visual indicators for processing status and results.

- **Bug Fixes**
  - Improved error handling for file upload and download failures, with clear user feedback.

- **Documentation**
  - Updated documentation and badges to correct spelling to British English and reflect new asynchronous file handling features.

- **Tests**
  - Added comprehensive tests for asynchronous file operations, including error scenarios and state management.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/cetmix/cetmix-tower?pullRequest=272) <br/>All committers have signed the CLA.

ivs-cetmix commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,52 @@
+        """
+        Trigger asynchronous download via job queue.
+        """
+        self.with_delay()._do_download(raise_error=raise_error)
```
Same here

Kurama2d replied:
Thank you for your comment, I’ve made the changes.

ivs-cetmix commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,52 @@
+    _inherit = "cx.tower.file"
+
+    is_processed = fields.Boolean(
+        string="Is Processed",
```
From the [OCA Coding Guidelines](https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#fields):
`
If the technical name of the field (the variable name) is the same to the string of the label, don't put string parameter for new API fields, because it's automatically taken. If your variable name contains "_" in the name, they are converted to spaces when creating the automatic string and each word is capitalized.
`

Kurama2d replied:
Thank you for your comment, I’ve made the changes.

ivs-cetmix commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,52 @@
+
+    is_processed = fields.Boolean(
+        string="Is Processed",
+        default=False,
```
`False` is the default value for the Boolean fields, no need to put this value explicitly.

Kurama2d replied:
Thank you for your comment, I’ve made the changes.

ivs-cetmix commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,52 @@
+    is_processed = fields.Boolean(
+        string="Is Processed",
+        default=False,
+        help="True if the file is already processed",
```
`True if the file is currently being processed `

Kurama2d replied:
thanks, reworked

ivs-cetmix commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,52 @@
+        """
+        Trigger asynchronous upload via job queue.
+        """
+        self.with_delay()._do_upload(raise_error=raise_error)
```
What if this file is already being processed? We should check this before we proceed.

Kurama2d replied:
Thank you for your comment, I’ve added a check for file processing.

ivs-cetmix commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,52 @@
+        Sets 'is_processed' to True before the upload and to False after.
+        """
+        self.ensure_one()
+        self.is_processed = True
```
We should set this flag outside this function because this method is queued. So we must ensure that no other task will be queued while this one is waiting.
Also this flag simply will not work because it's set and reset within the same transaction.

Kurama2d replied:
Thank you for your comment, I’ve moved setting the flag to True outside the method.

ivs-cetmix reviewed (changes requested):
Please remember that you should answer all the rabbit comments before submitting code for review.

Kurama2d commented on cetmix_tower_server_queue/static/description/index.html:
```diff
@@ -369,7 +369,7 @@ <h1 class="title">Cetmix Tower Server Queue</h1>
 !! source digest: sha256:def1b88b3682af158d70ee6e4b0b002869b73018fab92c4242728934cc009bd8
 !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! -->
-<p><a class="reference external image-reference" href="https://odoo-community.org/page/development-status"><img alt="Production/Stable" src="https://img.shields.io/badge/maturity-Production%2FStable-green.png" /></a> <a class="reference external image-reference" href="http://www.gnu.org/licenses/agpl-3.0-standalone.html"><img alt="License: AGPL-3" src="https://img.shields.io/badge/license-AGPL--3 ...
+<p><a class="reference external image-reference" href="https://odoo-community.org/page/development-status"><img alt="Production/Stable" src="https://img.shields.io/badge/maturity-Production%2FStable-green.png" /></a> <a class="reference external image-reference" href="http://www.gnu.org/licenses/agpl-3.0-standalone.html"><img alt="License: AGPL-3" src="https://img.shields.io/badge/licence-AGPL--3 ...
```
These changes are generated automatically.

Kurama2d commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,122 @@
+            # Verify flag is reset after operation
+            self.refresh()  # Refresh to get latest DB values
+            return result
+
```
I’ve updated the tests, the suggested edits are no longer needed.

Kurama2d commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,122 @@
+        # Apply our patches
+        type(self.file)._do_upload = patched_do_upload
+        type(self.file)._do_download = patched_do_download
+
```
I’ve updated the tests, the suggested edits are no longer needed.

Kurama2d commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,122 @@
+            self.write({"is_processed": True})
+            # Set a server response
+            self.server_response = "ok"
+            result = orig_do_upload(self, raise_error=raise_error)
```
I’ve updated the tests, the suggested edits are no longer needed.

Kurama2d commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,122 @@
+                self.file_2.server_response,
+                "ok",
+                "Server response should be 'ok' after upload",
+            )
```
I’ve updated the tests, the suggested edits are no longer needed.

Kurama2d commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,144 @@
+        error_file = self.File.create(
+            {
+                "source": "tower",
+                "template_id": self.file_template.id,
+                "server_id": self.server_test_1.id,
+                "is_processed": False,
+                "name": "error_test_file",  # Unique name for this test
+            }
+        )
+
+        # Mock with_delay to get a job object
+        with mock.patch(
+            "odoo.addons.queue_job.models.base.Base.with_delay"
+        ) as mock_with_delay:
+            mock_job = mock.MagicMock()
+            mock_with_delay.return_value = mock_job
+
+            # Trigger upload which will create a job (this should not raise an error)
+            error_file.upload()
+
+            # Verify _do_upload was requested with correct params
+            mock_job._do_upload.assert_called_once_with(raise_error=False)
+
+            # At this point, file would be marked as processed in a real scenario
+            error_file.write({"is_processed": True})
+            self.assertTrue(
+                error_file.is_processed, "Flag should be set during processing"
+            )
+
+            # Now simulate what happens when an error occurs during job execution
+
+            # Direct call to write without mocking - this actually updates the database
+            error_file.write({"is_processed": False})
+
+            # Verify the flag is reset after error handling
+            error_file.refresh()  # Ensure we have latest database state
+            self.assertFalse(
+                error_file.is_processed,
+                "is_processed flag should be reset after error handling",
+            )
```
Using a general Exception here might be too broad. In our case, _do_upload may need to distinguish between different types of exceptions, and catching all exceptions could mask unintended issues.
It would be better to mock and test with a specific, expected exception type if needed.

Kurama2d replied:
I agree with the changes, I’ve updated the test_upload_error_handling test.

ivs-cetmix commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,57 @@
+        """
+        self.ensure_one()
+        if self.is_processed:
+            raise UserError(_("This file is already being processed."))
```
This function has a `raise_error` argument, which should be considered every time an exception is raised.

Kurama2d replied:
Thank you for the feedback.
I am planning to check raise_error in the condition (e.g., if self.is_processed and raise_error:).
In case raise_error == False, should the method simply return silently, or is there a preferred way to handle this case (e.g., logging, warning, or a specific return value)?

ivs-cetmix replied:
Let's save response in this [this field](https://github.com/cetmix/cetmix-tower/blob/3a9e03d8d58df586472dcc0046aac0bf9687406d/cetmix_tower_server/models/cx_tower_file.py#L101).
In case  there is no response, let's add the following pre-defined error codes in the [constants](https://github.com/cetmix/cetmix-tower/blob/3a9e03d8d58df586472dcc0046aac0bf9687406d/cetmix_tower_server/models/constants.py#L55) and save them instead.

401 FILE_UPLOAD_FAILED
402 FILE_DOWNLOAD_FAILED

Kurama2d replied:
I have added handling for cases with raise_error, and the predefined error codes FILE_UPLOAD_FAILED and FILE_DOWNLOAD_FAILED are now used.

ivs-cetmix commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,57 @@
+class CxTowerFile(models.Model):
+    _inherit = "cx.tower.file"
+
+    is_processed = fields.Boolean(
```
I think we should rename this field into `is_being_processed` to avoid possible misunderstanding.

Kurama2d replied:
Absolutely agree, changed.

ivs-cetmix commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,57 @@
+        if self.is_processed:
+            raise UserError(_("This file is already being processed."))
+        self.is_processed = True
+        self.with_delay()._do_upload(raise_error=raise_error)
```
File upload operation can be also triggered by the [upload_file_by_template](https://github.com/cetmix/cetmix-tower/blob/47e7d6a5ab046ca111dfbc0dbe867842f6cb2cea/cetmix_tower_server/models/cx_tower_server.py#L1138) command which in its turn can be also [queued](https://github.com/cetmix/cetmix-tower/blob/47e7d6a5ab046ca111dfbc0dbe867842f6cb2cea/cetmix_tower_server_queue/models/cx_tower_server.py#L24).
We need to detect such case and avoid queueing an already queued operation.

Kurama2d replied:
Thank you for the clarification.
I have added a check to avoid queueing an already queued operation.

ivs-cetmix commented on cetmix_tower_server_queue/tests/test_file.py:
We need to ensure that the corresponding error codes are set when a file operation succeeds or fails.

Kurama2d replied:
I have updated the tests according to the changed logic.

ivs-cetmix reviewed (changes requested):
We

Kurama2d commented on cetmix_tower_server_queue/views/cx_tower_file_view.xml:
```diff
@@ -0,0 +1,64 @@
+                </attribute>
+                <attribute name="decoration-danger">
+                    is_being_processed == False and server_response != 'ok' and server_response != False
+                </attribute>
```
I updated the condition for the decoration-info attribute, but for consistency with the rest of the code, it's clearer to use direct conditions with == False.

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,86 @@
+class CxTowerFile(models.Model):
+    _inherit = "cx.tower.file"
+
+    is_being_processed = fields.Boolean(
+        help="True if the file is currently being processed",
+    )
```
I don't think that it's needed because it's an invisible field and can't be toggled manually,

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,86 @@
+        Perform the upload and update processing status.
+        Sets 'is_being_processed' to True before the upload and to False after.
+        """
+        self.ensure_one()
```
You're right, fixed.

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,86 @@
+        """
+        self.ensure_one()
+
+        # Check if the file is already being processed
+        if self.is_being_processed and raise_error:
+            raise UserError(_("This file is already being processed."))
+        elif self.is_being_processed and not raise_error:
+            self.server_response = FILE_UPLOAD_FAILED
+            return
```
It was considered to save server_response in this case.

Kurama2d replied:
In this case, we explicitly set the server_response to an error code to mark this transaction attempt.
Later, when the file processing is completed, the server_response will be updated according to the actual result based on the further processing logic.

Kurama2d commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,264 @@
+            "odoo.addons.cetmix_tower_server.models.cx_tower_file.CxTowerFile.upload",
+        ) as mock_upload:
+
+            def upload_and_raise(*args, **kwargs):
+                error_file.server_response = "error"
+                raise exceptions.ValidationError(_("Specific validation error"))
```
For this test I use "error" string as example of any string that can be in server_response.

Kurama2d commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,264 @@
+            "odoo.addons.cetmix_tower_server.models.cx_tower_file.CxTowerFile.download",
+        ) as mock_download:
+
+            def download_and_raise(*args, **kwargs):
+                error_file.server_response = "error"
+                raise exceptions.ValidationError(_("Specific validation error"))
```
For this test I use "error" string as example of any string that can be in server_response.

Kurama2d commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,264 @@
+            error_file.is_being_processed,
+            "is_being_processed should be False even after a validation error",
+        )
+        self.assertNotEqual(
+            error_file.server_response,
+            "ok",
+            "Server response should not be 'ok' after a validation error",
+        )
```
str(FILE_UPLOAD_FAILED) is not te same as "ok", so it's not relevant here

Kurama2d commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,264 @@
+            error_file.is_being_processed,
+            "is_being_processed should be False even after a download validation error",
+        )
+        self.assertNotEqual(
+            error_file.server_response,
+            "ok",
+            "Server response should not be 'ok' after a download validation error",
+        )
```
str(FILE_DOWNLOAD_FAILED) is not te same as "ok", so it's not relevant here

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,86 @@
+class CxTowerFile(models.Model):
+    _inherit = "cx.tower.file"
+
+    is_being_processed = fields.Boolean(
+        help="True if the file is currently being processed",
+    )
```
I don't think that it's needed because it's an invisible field and can't be toggled manually,

ivs-cetmix commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,86 @@
+
+        # Enqueue the upload if not already in a queue job;
+        # otherwise, execute immediately
+        if not self.env.context.get("job_uuid"):
```
Question, not an issue:  where do we get this value in the context from?

Kurama2d replied:
https://github.com/OCA/queue/blob/14.0/queue_job/job.py#L715
The job_uuid is automatically set in the context by the queue_job module when the job is executed, so that methods can detect they are running inside a queued job.

Here are examples of the context output in queued\not queued methods

![image](https://github.com/user-attachments/assets/0d0d52ae-8929-4f4e-82d9-557a7cfebc42)

![image](https://github.com/user-attachments/assets/572042c0-83c6-48f6-8a1a-be8a455f0056)

ivs-cetmix replied:
Perfect, thank you!

Kurama2d commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,264 @@
+            }
+        )
+
+        with mock.patch(
+            "odoo.addons.queue_job.models.base.Base.with_delay"
+        ) as mock_with_delay:
+
+            def make_upload_job(file_record):
+                job = mock.MagicMock()
+
+                def do_upload(raise_error=False):
+                    file_record.server_response = "ok"
+
+                job._do_upload.side_effect = do_upload
+                return job
+
+            mock_jobs = [
+                make_upload_job(upload_file),
+                make_upload_job(upload_file_2),
+            ]
+            mock_with_delay.side_effect = mock_jobs
```
Tests are done and no changes are needed, the file hasn't been changed.

Kurama2d commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,264 @@
+                "server_id": self.server_test_1.id,
+                "server_dir": "/var/tmp",
+                "is_being_processed": False,
+            }
+        )
+
+        with mock.patch(
+            "odoo.addons.queue_job.models.base.Base.with_delay"
+        ) as mock_with_delay:
+
+            def make_upload_job(file_record):
+                job = mock.MagicMock()
+
+                def do_upload(raise_error=False):
+                    file_record.server_response = "ok"
+
+                job._do_upload.side_effect = do_upload
+                return job
+
+            mock_jobs = [
+                make_upload_job(upload_file),
+                make_upload_job(upload_file_2),
+            ]
+            mock_with_delay.side_effect = mock_jobs
+
+            # Trigger uploads
+            upload_file.upload()
+            upload_file_2.upload()
+
+            self.assertEqual(
+                upload_file.server_response,
+                "ok",
+                "Server response should be 'ok' after upload",
+            )
+            self.assertEqual(
+                upload_file_2.server_response,
+                "ok",
+                "Server response should be 'ok' after upload",
+            )
+
```
Tests are done and no changes are needed, the file hasn't been changed.

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,122 @@
+        copy=False,
+    )
+
+    def _compute_is_being_processed(self):
+        """
+        Compute if the file is being processed based on job state.
+        """
+        for record in self:
+            record.is_being_processed = record.queue_job_state in [
+                "wait_dependencies",
+                "pending",
+                "enqueued",
+                "started",
+            ]
```
There is no `draft` state in `queue.job`

Kurama2d commented on cetmix_tower_server_queue/models/queue_job.py:
```diff
@@ -0,0 +1,18 @@
+        res = super(QueueJob, self).write(vals)
+        if "state" in vals and self.model_name == "cx.tower.file":
+            self.records.write({"queue_job_state": vals["state"]})
+        return res
```
agree, fixed

Kurama2d commented on cetmix_tower_server_queue/views/cx_tower_file_view.xml:
```diff
@@ -0,0 +1,58 @@
+                <field name="queue_job_state" invisible="1" />
+            </xpath>
+            <xpath expr="//tree" position="attributes">
+                <attribute name="decoration-info">
+                    is_being_processed == True
+                </attribute>
```
it's the same but shorter, so i decided to have is_being_processed here

Kurama2d commented on cetmix_tower_server_queue/views/cx_tower_file_view.xml:
```diff
@@ -0,0 +1,58 @@
+                    queue_job_state == 'done' and server_response == 'ok'
+                </attribute>
+                <attribute name="decoration-danger">
+                    server_response not in ('ok', False) or queue_job_state == 'failed'
+                </attribute>
```
I think that it's possible to have my option here

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,122 @@
+class CxTowerFile(models.Model):
+    _inherit = "cx.tower.file"
+
+    queue_job_state = fields.Char(
+        help="State of the last queue job for this file",
+    )
+    is_being_processed = fields.Boolean(
```
Thought about it, but decided to replace to use it for custom states in future

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,120 @@
+        Compute if the file is being processed based on job state.
+        """
+        for record in self:
+            record.is_being_processed = record.queue_job_state in [
+                "wait_dependencies",
+                "pending",
+                "enqueued",
+                "started",
+            ]
```
this approach saves space for future improvements

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,122 @@
+            self.with_delay()._do_upload(raise_error=raise_error)
+        else:
+            self._do_upload(raise_error=raise_error)
+
```
removed ensure one and adapted to multi-records

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,122 @@
+                raise e
+            else:
+                _logger.error("File %s upload failed: %s", str(self), str(e))
+
```
fixed

Kurama2d commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,271 @@
+                # Simulate what happens after an error in the actual implementation
+                error_file.server_response = "error"
+                raise exceptions.ValidationError(_("Specific validation error"))
+
```
fixed

Kurama2d commented on cetmix_tower_server_queue/views/cx_tower_file_view.xml:
```diff
@@ -0,0 +1,58 @@
+                <field name="queue_job_state" invisible="1" />
+            </xpath>
+            <xpath expr="//tree" position="attributes">
+                <attribute name="decoration-info">
+                    is_being_processed == True
+                </attribute>
+                <attribute name="decoration-success">
+                    queue_job_state == 'done' or server_response == 'ok'
+                </attribute>
+                <attribute name="decoration-danger">
+                    server_response not in ('ok', False) or queue_job_state == 'failed'
+                </attribute>
```
There are states of job included in is_being_processed logic, so there is no option when queue_job_state == 'failed' and states of processing come at the same time.

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,176 @@
+                "params": {
+                    "title": _("Failure"),
+                    "message": _(
+                        "Unable to upload file '%(f)s'.\n"
+                        "Upload operation is not supported for 'server' type files.",
+                        f=server_files[0].rendered_name,
+                    ),
```
It's existing inherited code, I won't change this one.

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,172 @@
+                    },
+                )
+                return result
+        except Exception as e:
```
Different messages for one and multi records are added.

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,172 @@
+        """
+        Compute if the file is being processed based on job state.
+        """
+        for record in self:
+            record.is_being_processed = record.queue_job_state in [
+                "wait_dependencies",
+                "pending",
+                "enqueued",
+                "started",
+            ]
```
In this case let's keep this option.

Kurama2d commented on cetmix_tower_server_queue/models/queue_job.py:
```diff
@@ -0,0 +1,20 @@
+        if "state" in vals:
+            for job in self:
+                if self.model_name == "cx.tower.file":
+                    job.records.write({"queue_job_state": vals["state"]})
```
Agree, fixed

Kurama2d commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,271 @@
+            "odoo.addons.cetmix_tower_server.models.cx_tower_file.CxTowerFile.download",
+        ) as mock_download:
+
+            def download_and_raise(*args, **kwargs):
+                error_file.server_response = "error"
+                raise exceptions.ValidationError(_("Specific validation error"))
```
It can be saved for test.

GabbasovDinar replied:
@Kurama2d we have already patched download method: https://github.com/cetmix/cetmix-tower/blob/14.0/cetmix_tower_server/tests/common.py#L311

I suggest you expand this method in cetmix_tower_server to raise error: maybe via context (or otherwise), e.g.:

```python

...
error_file.with_context(raise_download_error="Download Error")._do_download(raise_error=True)
...

def ssh_download_file(self, remote_path):
    error = self.env.context.get("raise_download_error")
    if error:
        raise exceptions.ValidationError(error)

    _, extension = os.path.splitext(remote_path)
    if extension == ".zip":
        return b"ok\x00"
    return b"ok"

```

what do you think?

Kurama2d replied:
thanks, used it

Kurama2d commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,271 @@
+
+        # In actual queue_job execution, the job would be marked as failed
+        # We simulate this behavior for the test
+        error_file.write(
+            {"queue_job_state": "failed", "server_response": FILE_UPLOAD_FAILED}
+        )
+        error_file.refresh()
+
+        # Verify the file state after error
+        self.assertEqual(
+            error_file.queue_job_state,
```
Just refresh() is not enough.

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,120 @@
+    def upload(self, raise_error=False):
+        """
+        Trigger asynchronous upload via job queue.
+        """
+        # Check if the file is already being processed
+        if any(self.mapped("is_being_processed")):
+            if raise_error:
+                raise UserError(_("This file is already being processed."))
+            else:
+                self.write({"server_response": FILE_UPLOAD_FAILED})
+                return
+
```
It makes no difference, if the transaction is aborted, the results of file processing will be rolled back.

Kurama2d commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,271 @@
+            {"queue_job_state": "failed", "server_response": FILE_UPLOAD_FAILED}
+        )
+        error_file.refresh()  # Ensure we have fresh data from DB
+
```
discussed below

Kurama2d replied:
this part of code is changed

Kurama2d commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,271 @@
+            "odoo.addons.queue_job.models.base.Base.with_delay"
+        ) as mock_with_delay:
+            # Set up mock jobs that will be returned by with_delay()
+            mock_jobs = []
+
+            for file_record in [upload_file, upload_file_2]:
+                job = mock.MagicMock()
+
+                # Create a side effect that simulates successful upload
+                def do_upload(file_obj):
+                    def _do_upload_impl(*args, **kwargs):
+                        file_obj.server_response = "ok"
+
+                    return _do_upload_impl
+
+                # Set the side effect for this specific job
+                job._do_upload.side_effect = do_upload(file_record)
+                mock_jobs.append(job)
+
+            mock_with_delay.side_effect = mock_jobs
+
```
not relevant anymore

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,151 @@
+                raise
+            else:
+                self.write({"is_being_processed": False})
+                _logger.error("File %s download failed: %s", str(self), str(e))
```
server_response will be filled during upload or download processing, we don't need to set it manually

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,151 @@
+        """
+        Trigger asynchronous upload via job queue.
+        """
+        # Check if the file is already being processed
+        if any(self.mapped("is_being_processed")):
+            if raise_error:
+                raise UserError(_("This file is already being processed."))
+            else:
+                return
+
```
We shouldn't set server_response on repeated method calls, as it may overwrite the original error from the actual job execution with an error from a later call.

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,151 @@
+        Uploads the file within a job context and notifies the user on success.
+        Sets failure state and logs the error if an exception occurs.
+        """
+        try:
+            with self.env.cr.savepoint():
+                result = super().upload(raise_error=raise_error)
+                single_msg = _("File uploaded!")
+                plural_msg = _("Files uploaded!")
+                self.env["bus.bus"].sudo().sendone(
```
there is no ensure_one() in base module upload or download methods

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,151 @@
+        Downloads the file within a job context and notifies the user on success.
+        Sets failure state and logs the error if an exception occurs.
+        """
+        try:
+            with self.env.cr.savepoint():
+                result = super().download(raise_error=raise_error)
+                single_msg = _("File downloaded!")
+                plural_msg = _("Files downloaded!")
+                self.env["bus.bus"].sudo().sendone(
```
there is no ensure_one() in base module upload or download methods

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,151 @@
+                )
+                self.write({"is_being_processed": False})
+                return result
+        except Exception as e:
+            if raise_error:
+                raise
+            else:
+                self.write({"is_being_processed": False})
+                _logger.error("File %s upload failed: %s", str(self), str(e))
+
```
We shouldn't set server_response on repeated method calls, as it may overwrite the original error from the actual job execution with an error from a later call.

Kurama2d committed:
[IMP] cetmix_tower_server_queue: async file operations

Use queue_job to process file upload/download in background.
Add is_being_processed field and UI indicators for status display.

Task: 3720

Aldeigja reviewed (approved):
Functional LGTM

GabbasovDinar commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,214 @@
+            }
+        )
+
+        with mock.patch(
+            "odoo.addons.queue_job.models.base.Base.with_delay"
+        ) as mock_with_delay:
+            # Set up mock jobs that will be returned by with_delay()
+            mock_jobs = []
+
+            for file_record in [download_file, download_file_2]:
+                job = mock.MagicMock()
+
+                # Create a side effect that simulates successful download
+                def do_download(file_obj):
+                    def _do_download_impl(*args, **kwargs):
+                        file_obj.server_response = "ok"
+                        file_obj.is_being_processed = False
+
+                    return _do_download_impl
+
+                # Set the side effect for this specific job
+                job._do_download.side_effect = do_download(file_record)
+                mock_jobs.append(job)
+
+            mock_with_delay.side_effect = mock_jobs
```
queue_job already mocked in tests, please use mocked process.

see documentation: https://github.com/OCA/queue/tree/14.0/queue_job#testing

example:
https://github.com/cetmix/cetmix-tower/pull/284/files#diff-21009aec6edfd309afab498e23fb71de5cad592e8e8596ffafbcfb9a9558281fR112

Kurama2d replied:
I’ve encountered a few issues when trying to use trap_jobs():

I couldn’t find working examples of using trap_jobs() in tests outside the queue_job module itself. I followed the documentation, but I’m getting an error from within queue_job:
RuntimeError: 'auto_spec' might be a typo... — this happens when entering the trap_jobs() context.

I tried running the cetmix_tower_server_queue module locally using the branch 14.0-t4264-cetmix_tower_queue-run-child-flight-plan-commands-one-by-one, and I get the same error when using trap_jobs().

I'm not sure it's appropriate for me to modify the common.py file, as it could affect other tests or modules.

I also noticed that the [documentation](https://github.com/OCA/queue/tree/14.0/queue_job#testing) refers to odoo.addons.queue_job.tests.common.JobsTester, but this class doesn’t seem to exist — most likely the documentation is outdated or inaccurate.

At the moment, my current implementation using mocks is working and doesn't block the task.
If there’s a more detailed and up-to-date example of using trap_jobs() correctly (especially with a workaround for the current error), I’d be happy to apply it. Otherwise, I’d prefer to keep the current approach to avoid delays.

Thanks!

GabbasovDinar replied:
@Kurama2d 
Try to change `auto_spec` to `autospec` in queue_job module in your local repository, and re-run tests:

```
index 5ca2f079..45fd20de 100644
--- a/queue_job/tests/common.py
+++ b/queue_job/tests/common.py
@@ -95,7 +95,7 @@ def trap_jobs():
with mock.patch(
"odoo.addons.queue_job.delay.Job",
name="Job Class",
- auto_spec=True,
+ autospec=True,
) as job_cls_mock:
with JobsTrap(job_cls_mock) as trap:
yield trap
```

Kurama2d replied:
Thanks for clarifying, fixed it
autospec helped to pass tests localy

GabbasovDinar commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,214 @@
+            }
+        )
+
+        with mock.patch(
```
queue_job already mocked in tests, please use mocked process.

Kurama2d replied:
Please check [this reply](https://github.com/cetmix/cetmix-tower/pull/272#discussion_r2091980850)

Kurama2d replied:
Thanks for clarifying, fixed it

GabbasovDinar commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,214 @@
+
+        # Mock super().upload to raise a specific exception type
+        with mock.patch(
+            "odoo.addons.cetmix_tower_server.models.cx_tower_file.CxTowerFile.upload",
```
Please try using the same method I suggested to you for downloading

Kurama2d replied:
Thanks for suggestion, but I decided to not expand this method in cetmix_tower_server, because of some errors raised in some tests.

GabbasovDinar replied:
@Kurama2d What kind of mistakes? In any case I don't really understand the need to create a patch when there is already a patch for this method. and it is enough to use it, so there is no need to create the same patch in tests in different places.

Kurama2d replied:
Thanks for clarifying, fixed it

GabbasovDinar commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,150 @@
+        # Check if the file is already being processed
+        if any(self.mapped("is_being_processed")):
+            if raise_error:
+                raise UserError(_("This file is already being processed."))
```
Perhaps we need to add more detail to clarify which file was processed e.g.:

```python

processed_files = self.filtered(lambda file: file.is_being_processed)
if processed_files:
    raise UserError(_("The following files have already been processed: %(name)s", name=",".join(processed_files.mapped("name"))))

```

Kurama2d replied:
Thanks! Totally agree, added details to the message

GabbasovDinar commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,150 @@
+    )
+
+    server_response = fields.Text(
+        copy=False,
```
I think we can set this attribute in `cetmix_tower_server` module

Kurama2d replied:
Agree, done

GabbasovDinar commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,150 @@
+        """
+
+        # Check if the file is already being processed
+        if any(self.mapped("is_being_processed")):
+            if raise_error:
+                raise UserError(_("This file is already being processed."))
+            else:
+                return
```
please, add more file details in error message

Kurama2d replied:
Thanks! Totally agree, added details to the message

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,150 @@
+            else:
+                return
+
+        self.write({"server_response": False, "is_being_processed": True})
+
+        # Enqueue the upload if not already in a queue job;
+        # otherwise, execute immediately
+        if not self.env.context.get("job_uuid"):
+            self.with_delay()._do_upload(raise_error=raise_error)
+        else:
+            self._do_upload(raise_error=raise_error)
```
it's not possible to enqueue a duplicate job because we check if file is being processed, not job state

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,158 @@
+                self.write({"is_being_processed": False})
+                _logger.error("File %s upload failed: %s", str(self), str(e))
+
+    def _do_download(self, raise_error=True):
+        """
+        Downloads the file within a job context and notifies the user on success.
+        Sets failure state and logs the error if an exception occurs.
+        """
+        try:
+            with self.env.cr.savepoint():
+                result = super().download(raise_error=raise_error)
+                single_msg = _("File downloaded!")
+                plural_msg = _("Files downloaded!")
+                self.env["bus.bus"].sudo().sendone(
+                    (self._cr.dbname, "res.partner", self.env.user.partner_id.id),
+                    {
+                        "message": single_msg if len(self) == 1 else plural_msg,
+                        "title": _("Success"),
+                        "type": "success",
+                    },
+                )
+                self.write({"is_being_processed": False})
+                return result
+        except Exception as e:
+            if raise_error:
+                raise
+            else:
+                self.write({"is_being_processed": False})
+                _logger.error("File %s download failed: %s", str(self), str(e))
+
```
this logic hasn't been changed in last commit

Kurama2d replied:
it's done

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,158 @@
+                )
+            else:
+                return
+
```
this logic hasn't been changed in last commit

Kurama2d replied:
I prefer to keep the code as is

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,158 @@
+class CxTowerFile(models.Model):
+    _inherit = "cx.tower.file"
+
+    is_being_processed = fields.Boolean(
+        copy=False,
+        help="True if the file is currently being processed",
+    )
```
this logic hasn't been changed in last commit

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,158 @@
+        try:
+            with self.env.cr.savepoint():
+                result = super().download(raise_error=raise_error)
+                single_msg = _("File downloaded!")
+                plural_msg = _("Files downloaded!")
+                self.env["bus.bus"].sudo().sendone(
+                    (self._cr.dbname, "res.partner", self.env.user.partner_id.id),
+                    {
+                        "message": single_msg if len(self) == 1 else plural_msg,
+                        "title": _("Success"),
+                        "type": "success",
+                    },
```
bus.bus.sendone() pushes are small and fast, so they don't need to be so informative
and this logic hasn't been changed in last commit

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,148 @@
+                result = super().upload(raise_error=raise_error)
+                single_msg = _("File uploaded!")
+                plural_msg = _("Files uploaded!")
+                self.env["bus.bus"].sudo().sendone(
+                    (self._cr.dbname, "res.partner", self.env.user.partner_id.id),
+                    {
+                        "message": single_msg if len(self) == 1 else plural_msg,
+                        "title": _("Success"),
+                        "type": "success",
+                    },
+                )
+                self.write({"is_being_processed": False})
+                return result
+        except Exception as e:
+            if raise_error:
+                raise
+            else:
+                self.write({"is_being_processed": False})
+                _logger.error("File %s upload failed: %s", str(self), str(e))
```
this logic hasn't been changed in last commit

Kurama2d replied:
this part is already changed

GabbasovDinar reviewed (changes requested):
check my comments

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,157 @@
+                    },
+                )
+                return result
+        except Exception as e:
+            if not raise_error:
+                _logger.error("File %s upload failed: %s", str(self), str(e))
+            else:
+                raise
+        finally:
+            self.write({"is_being_processed": False})
```
It's not needed to log when we have raise_error=True because we have exception raised

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,158 @@
+                raise
+        finally:
+            self.write({"is_being_processed": False})
+
```
Fixed

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,158 @@
+                raise
+        finally:
+            self.write({"is_being_processed": False})
+
```
fixed

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,158 @@
+
+    is_being_processed = fields.Boolean(
+        help="True if the file is currently being processed",
+    )
```
saved copy=False to prevent leaking transient state when duplicating records

GabbasovDinar commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,157 @@
+        Trigger asynchronous upload via job queue.
+        """
+        # Check if the file is already being processed
+        if self._abort_if_processing(raise_error):
```
self._check_processed_files(raise_error)

Kurama2d replied:
thanks, renamed

GabbasovDinar commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,157 @@
+        """
+
+        # Check if the file is already being processed
+        if self._abort_if_processing(raise_error):
```
self._check_processed_files(raise_error)

Kurama2d replied:
thanks, renamed

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,151 @@
+            else:
+                # self.write({"server_response": FILE_DOWNLOAD_FAILED})
+                return
+
```
outdated comment, issue resolved and commented

Kurama2d commented on cetmix_tower_server_queue/views/cx_tower_file_view.xml:
```diff
@@ -0,0 +1,56 @@
+                <field name="server_response" invisible="1" />
+            </xpath>
+            <xpath expr="//tree" position="attributes">
+                <attribute name="decoration-info">
+                    is_being_processed == True
+                </attribute>
```
outdated comment, issue resolved and commented

ivs-cetmix commented:
@Kurama2d please ensure that all comments are answered.

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,157 @@
+        help="True if the file is currently being processed",
+    )
+
+    def _check_processed_files(self, raise_error):
+        """
+        Check if any file in the recordset is being processed.
+        If so, raise an error or return without doing anything.
+        """
```
it returns True as a sign of silent work

ivs-cetmix replied:
Please update the docstring.

Kurama2d replied:
Thanks, fixed docstring

ivs-cetmix commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,157 @@
+
+    is_being_processed = fields.Boolean(
+        copy=False,
+        help="True if the file is currently being processed",
```
Opt for using less technical statements in UI because it's meant for non technical people.
`help="File is currently being processed"`

Kurama2d replied:
Thanks, agree, fixed it

ivs-cetmix reviewed (changes requested):
All comments must be answered before PR is submitted for review.

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,158 @@
+                )
+            else:
+                return
+
```
fixed

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,158 @@
+                )
+            else:
+                return
+
```
fixed

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,148 @@
+                raise UserError(_("The following files have already been processed: %(name)s", name=",".join(processed_files.mapped("name"))))
+            else:
+                return
+
```
fixed

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,148 @@
+                raise UserError(_("The following files have already been processed: %(name)s", name=",".join(processed_files.mapped("name"))))
+            else:
+                return
+
```
fixed

Kurama2d commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,271 @@
+                        file_obj.server_response = "ok"
+
+                    return _do_upload_impl
+
```
I prefer to keep the code as is

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,122 @@
+            if raise_error:
+                raise e
+            else:
+                _logger.error("File %s download failed: %s", (str(self), str(e)))
```
fixed

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,122 @@
+            if raise_error:
+                raise e
+            else:
+                _logger.error("File %s upload failed: %s", (str(self), str(e)))
```
fixed

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,122 @@
+        try:
+            with self.env.cr.savepoint():
+                return super().upload(raise_error=raise_error, a=1)
+        except Exception as e:
```
fixed

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,122 @@
+        try:
+            with self.env.cr.savepoint():
+                return super().download(raise_error=raise_error, a=1)
+        except Exception as e:
```
fixed

Kurama2d commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,127 @@
+        self.assertFalse(
+            error_file.is_processed,
+            "is_processed should be False even after a validation error",
+        )
```
Added

Kurama2d commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,127 @@
+            self.assertFalse(
+                error_file.is_processed,
+                "is_processed should be False even after a validation error",
+            )
```
These tests are reworked

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,48 @@
+        result = super().upload(raise_error=raise_error)
+        self.is_processed = False
+        return result
+
```
this block is reworked

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,48 @@
+    def _do_download(self, raise_error=False):
+        """
+        Download method called with delay
+        Sets 'is_processed' to True before the upload and to False after.
```
this method was changed

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,48 @@
+        self.is_processed = True
+        result = super().download(raise_error=raise_error)
+        self.is_processed = False
+        return result
```
this method was changed

Kurama2d commented on cetmix_tower_server_queue/views/cx_tower_file_view.xml:
```diff
@@ -0,0 +1,64 @@
+                <attribute name="decoration-danger">
+                    server_response != 'ok'
+                </attribute>
+            </xpath>
```
these conditions were changed

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,172 @@
+        """
+        Perform upload within the job context and reset the processing flag.
+        Assumes `is_being_processed` was set to True by the caller.
+        """
```
this method was changed

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,172 @@
+                self.write({"server_response": FILE_UPLOAD_FAILED})
+                return
+
+        self.write({"server_response": False, "queue_job_state": False})
```
this method was changed, there is no more queue_job_state

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,157 @@
+        Sets failure state and logs the error if an exception occurs.
+        """
+        try:
+            with self.env.cr.savepoint():
+                result = super().upload(raise_error=raise_error)
+                single_msg = _("File uploaded!")
+                plural_msg = _("Files uploaded!")
+                self.env["bus.bus"].sudo().sendone(
+                    (self._cr.dbname, "res.partner", self.env.user.partner_id.id),
+                    {
+                        "message": single_msg if len(self) == 1 else plural_msg,
+                        "title": _("Success"),
+                        "type": "success",
+                    },
+                )
+                return result
```
server_response is enough to see that process failed

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,157 @@
+        Check if any file in the recordset is being processed.
+        If so, raise an error or return without doing anything.
+        """
+        processing_files = self.filtered(lambda file: file.is_being_processed)
+        if processing_files:
+            if raise_error:
+                raise UserError(
+                    _(
+                        "The following files are already being processed: %(name)s",
+                        name=", ".join(processing_files.mapped("name")),
+                    )
+                )
+            else:
+                return True
+
```
It returns true to continue working silenltly

Kurama2d replied:
I meant aborting the operation silently when raise_error=False

ivs-cetmix commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,158 @@
+        help="File is currently being processed",
+    )
+
+    def _check_processed_files(self, raise_error):
```
I would suggest a more relevant name, eg `_files_beeing_processed`

Kurama2d replied:
Thanks! As per OCA conventions and general coding practices, method names should include a verb. I believe _check_processed_files is appropriate, but I can rename it if necessary.

ivs-cetmix replied:
@Kurama2d yes, you are right. I think we can opt for `_check_files_being_processed`

Kurama2d replied:
Agree, It's better, renamed

ivs-cetmix reviewed (changes requested):
Some minor updates.

Kurama2d commented on cetmix_tower_server_queue/readme/newsfragments/3720.feature:
```diff
@@ -0,0 +1 @@
+Upload and download files asynchronously.
```
I see, fixed it

Kurama2d commented on cetmix_tower_server/readme/newsfragments/3720.feature:
```diff
@@ -0,0 +1 @@
+Upload and download files asynchronously.
```
Thanks, fixed it

Kurama2d commented on cetmix_tower/readme/newsfragments/3720.feature:
```diff
@@ -0,0 +1 @@
+**cetmix_tower_server_queue**: Support asynchronous upload and download of files via the job queue.
```
Thanks, fixed it

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,158 @@
+        """
+        Push the file to server without success notification.
+        """
+        server_files = self.filtered(lambda file_: file_.source == "server")
+        if server_files:
+            return {
+                "type": "ir.actions.client",
+                "tag": "display_notification",
+                "params": {
+                    "title": _("Failure"),
+                    "message": _(
+                        "Unable to upload file '%(f)s'.\n"
+                        "Upload operation is not supported for 'server' type files.",
+                        f=server_files[0].rendered_name,
+                    ),
+                    "sticky": False,
+                },
+            }
```
This part was just taken from original method, no changes here are required

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,158 @@
+                    },
+                )
+                return result
+        except Exception as e:
+            if not raise_error:
+                _logger.error("File %s upload failed: %s", str(self), str(e))
+            else:
+                raise
+        finally:
```
I'd prefer to keep logger.error — it's a lighter alternative to raising an exception and keeps logs cleaner when processing many files. A short message feels more consistent in this context.

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,158 @@
+                        "type": "success",
+                    },
+                )
+                return result
+        except Exception as e:
+            if not raise_error:
+                _logger.error("File %s download failed: %s", str(self), str(e))
+            else:
+                raise
```
I'd prefer to keep logger.error — it's a lighter alternative to raising an exception and keeps logs cleaner when processing many files. A short message feels more consistent in this context.

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,159 @@
+        Check if any file in the recordset is being processed.
+        Raise an error if raise_error is True.
+        If raise_error is False, return True to signal the caller to exit silently.
+        """
```
Thanks, changed docstring

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,159 @@
+                    )
+                )
+            else:
+                return True
```
thanks, changed

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,160 @@
+        """
+        Uploads the file within a job context and notifies the user on success.
+        Sets failure state and logs the error if an exception occurs.
+        """
```
Thanks, changed docstrings

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,162 @@
+        """
+        Pull file from server without notification.
+        """
+        tower_files = self.filtered(lambda file_: file_.source == "tower")
+        server_files = self - tower_files
+
+        tower_files.action_get_current_server_code()
```
This methd was inherided and I prefer not to change anything inside.

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,162 @@
+
+        # Enqueue the upload if not already in a queue job;
+        # otherwise, execute immediately
+        if not self.env.context.get("job_uuid"):
+            self.with_delay()._do_upload(raise_error=raise_error)
+        else:
+            self._do_upload(raise_error=raise_error)
```
Thanks! I think this might be overkill — the current behavior works as expected in our context.

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,158 @@
+                )
+            else:
+                return True
+
```
Thanks, added return False

ivs-cetmix reviewed (approved):
Code review LGTM

CetmixGitDrone commented:
This PR has the `approved` label and has been created more than 5 days ago. It should therefore be ready to merge by a maintainer (or a PSC member if the concerned addon has no declared maintainer). 🤖

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,162 @@
+        help="File is currently being processed",
+    )
+
+    def _check_files_being_processed(self, raise_error):
+        """
+        Check if any file in the recordset is being processed.
+        True if at least one file is already processing and raise_error is False.
+        False if no files are currently being processed.
+        The caller uses the boolean to decide whether to continue or abort.
+        """
+        processing_files = self.filtered(lambda rec: rec.is_being_processed)
+        if processing_files:
+            if raise_error:
+                raise UserError(
+                    _(
+                        "The following files are already being processed: %(name)s",
+                        name=", ".join(processing_files.mapped("name")),
+                    )
+                )
+            else:
+                return True
+        return False
```
Let's save as it is

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,162 @@
+                result = super().upload(raise_error=raise_error)
+                single_msg = _("File uploaded!")
+                plural_msg = _("Files uploaded!")
+                self.env["bus.bus"].sudo().sendone(
+                    (self._cr.dbname, "res.partner", self.env.user.partner_id.id),
+                    {
+                        "message": single_msg if len(self) == 1 else plural_msg,
+                        "title": _("Success"),
+                        "type": "success",
+                    },
+                )
```
I think this might be overkill — the current behavior works as expected in our context.

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,162 @@
+
+        server_files.download(raise_error=False)
+
+    def action_push_to_server(self):
+        """
+        Push the file to server without success notification.
+        """
+        server_files = self.filtered(lambda file_: file_.source == "server")
+        if server_files:
+            return {
+                "type": "ir.actions.client",
+                "tag": "display_notification",
+                "params": {
+                    "title": _("Failure"),
+                    "message": _(
+                        "Unable to upload file '%(f)s'.\n"
+                        "Upload operation is not supported for 'server' type files.",
+                        f=server_files[0].rendered_name,
+                    ),
+                    "sticky": False,
+                },
+            }
```
It's inherited method, no changes here needed

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,162 @@
+        if self._check_files_being_processed(raise_error):
+            return
+
+        self.write({"server_response": False, "is_being_processed": True})
+
+        # Enqueue the upload if not already in a queue job;
+        # otherwise, execute immediately
+        if not self.env.context.get("job_uuid"):
+            self.with_delay()._do_upload(raise_error=raise_error)
+        else:
+            self._do_upload(raise_error=raise_error)
```
I think this might be overkill — the current behavior works as expected in our context.

ivs-cetmix commented:
/ocabot merge minor

CetmixGitDrone commented:
What a great day to merge this nice PR. Let's do it!
Prepared branch [14.0-ocabot-merge-pr-272-by-ivs-cetmix-bump-minor](https://github.com/cetmix/cetmix-tower/commits/14.0-ocabot-merge-pr-272-by-ivs-cetmix-bump-minor), awaiting test results.

CetmixGitDrone commented:
Congratulations, your PR was merged at 2eb2c55482ab0d6a7815973c6eaf7c5523f8440b. Thanks a lot for contributing to cetmix. ❤️
